### PR TITLE
Feature: business logic and entrypoints to get categories (21102023)

### PR DESCRIPTION
**Pull request by Azim-Azizov**

### CODE REVIEW

File: src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package me.svdev.mscategory.mapper;
+
+import me.svdev.mscategory.dao.entity.Category;
+import me.svdev.mscategory.model.dto.CategoryDto;
+import me.svdev.mscategory.model.request.CategoryRequest;
+
+public class CategoryMapper {
```
Comment (alihuseynaliyev) [inline src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java]: mapper-ləri class yox enuma dəyiş

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    public CategoriesResponse getCategories() {
+        return new CategoriesResponse(categoryRepository.findAll().stream().map(CategoryMapper::mapEntityToDto).toList());
+    }
+
+    public CategoryDto getCategory(Long id) {
+        return mapEntityToDto(getCategoryIfExists(id));
+    }
+
+//    public void saveCategory(CategoryRequest categoryRequest) {
```
Comment (alihuseynaliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: delete comments

File: src/test/java/me/svdev/mscategory/MsCategoryApplicationTests.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package me.svdev.mscategory;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class MsCategoryApplicationTests {
```
Comment (alihuseynaliyev) [inline src/test/java/me/svdev/mscategory/MsCategoryApplicationTests.java]: delete this test class and use groovy (spock) for testing

**Comment by ilkin-aliyev** [comment]
add .gitignore

File: src/main/java/me/svdev/mscategory/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @GetMapping("{id}")
+    public CategoryDto getCategory(@PathVariable Long id) {
+        return categoryService.getCategory(id);
+    }
+
+//    @PostMapping
+//    @ResponseStatus(CREATED)
+//    public void saveCategory(@RequestBody CategoryRequest categoryRequest) {
+//        categoryService.saveCategory(categoryRequest);
+//    }
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/controller/CategoryController.java]: remove this shit

File: src/main/java/me/svdev/mscategory/controller/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("v1/categories")
+public class CategoryController {
+
+    private final CategoryService categoryService;
+
+    @GetMapping
+    public CategoriesResponse categoriesResponse() {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/controller/CategoryController.java]: Rename to getCategories

File: src/main/java/me/svdev/mscategory/dao/entity/Category.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+@Getter
+@Setter
+@ToString
+@Entity
+@Table(name = "categories")
+public class Category {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(nullable = false, unique = true)
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/entity/Category.java]: don't need @Column validations in Entity class would be better use this logic in liquibase

File: src/main/java/me/svdev/mscategory/dao/entity/Category.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    private Long id;
+
+    @Column(nullable = false, unique = true, length = 32)
+    private String name;
+
+    @Column(nullable = false)
+    private Boolean isVisible = true;
+
+    @Column(nullable = false)
+    private Boolean isFavorite;
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/entity/Category.java]: If you write isVisible = true; you need to write also isFavorite=false, but i recommended use default values in liquibase and add this settings to mapper

File: src/main/java/me/svdev/mscategory/dao/repository/CategoryRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package me.svdev.mscategory.dao.repository;
+
+import me.svdev.mscategory.dao.entity.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/repository/CategoryRepository.java]: [Optional] - You don't need everything from JpaRepository you can use also CrudRepository

File: src/main/java/me/svdev/mscategory/dao/entity/Category.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+@ToString
+@Entity
+@Table(name = "categories")
+public class Category {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/entity/Category.java]: Use our naming convetions for classes rename to CategoryEntity

File: src/main/java/me/svdev/mscategory/model/response/CategoriesResponse.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import me.svdev.mscategory.model.dto.CategoryDto;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+public class CategoriesResponse {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/model/response/CategoriesResponse.java]: [Optional] - if you return 1 DS you can return it directly without creating any wrapper classes

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import static me.svdev.mscategory.mapper.CategoryMapper.mapEntityToDto;
+
+@RequiredArgsConstructor
+@Service
+public class CategoryService {
+
+    private final CategoryRepository categoryRepository;
+
+    public CategoriesResponse getCategories() {
+        return new CategoriesResponse(categoryRepository.findAll().stream().map(CategoryMapper::mapEntityToDto).toList());
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: if your category isVisible=false, you return this. this is bug, please don't return isVisible = false categories

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+//        var category = Category.builder()
+//                .name(categoryRequest.getName())
+//                .isVisible(true)
+//                .isFavorite(categoryRequest.getIsFavorite())
+//                .build();
+//        categoryRepository.save(category);
+//    }
+
+    private Category getCategoryIfExists(Long id) {
+        return categoryRepository.findById(id).orElseThrow(RuntimeException::new);
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: bad practice to throw exceptions from Java, create your custom exception, for example NotFoundException and add message to exception. Also you will be need error handling

File: src/main/resources/liquibase/changelogs/2023/10/19-01.yaml [hunk 13]
```diff
@@ -0,0 +1,49 @@
+databaseChangeLog:
+  - changeSet:
+      id: 1697693652731-1
```
Comment (ilkin-aliyev) [inline src/main/resources/liquibase/changelogs/2023/10/19-01.yaml]: would be better use more understandable id in here for example: Created categories table (also you can add time)

File: src/main/resources/liquibase/changelogs/2023/10/19-01.yaml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+                  constraints:
+                    nullable: false
+                    primaryKey: true
+                  name: id
+                  type: BIGINT
+              - column:
+                  constraints:
+                    nullable: false
+                  name: name
+                  type: VARCHAR(32)
```
Comment (ilkin-aliyev) [inline src/main/resources/liquibase/changelogs/2023/10/19-01.yaml]: 32 chars not enough for name would be better use TEXT

File: src/main/resources/liquibase/changelogs/2023/10/19-01.yaml [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+              - column:
+                  constraints:
+                    nullable: false
+                  name: is_visible
+                  type: BOOLEAN
+                  defaultValueBoolean: true
+              - column:
+                  constraints:
+                    nullable: false
+                  name: is_favorite
```
Comment (ilkin-aliyev) [inline src/main/resources/liquibase/changelogs/2023/10/19-01.yaml:30]: add default

File: src/main/resources/liquibase/changelogs/2023/10/19-01.yaml [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+              - column:
+                  name: updated_at
+                  type: TIMESTAMP
+                  defaultValue: now()
+              - column:
+                  name: created_at
+                  type: TIMESTAMP
+                  defaultValue: now()
+            tableName: categories
+  - changeSet:
```
Comment (ilkin-aliyev) [inline src/main/resources/liquibase/changelogs/2023/10/19-01.yaml]: you don't need second changeSet, you can add unique in 19 line

**Comment by ilkin-aliyev** [comment]
Logic not ended, because not exist SubCategory logic and one to many relation

File: src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java (hunk 1 above)
Comment (Azim-Azizov) [reply src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java]: Necə ki? class sözünü silib enum yazdım, errorlar axışmağa başladı)

**Comment by Azim-Azizov** [comment]
> add .gitignore

There was an actual .gitignore file)

File: .jpb/jpb-settings.xml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ilkin-aliyev) [inline .jpb/jpb-settings.xml]: remove this file

File: build.gradle [hunk 18]
```diff
@@ -0,0 +1,41 @@
+    plugins {
+    id 'java'
+    id 'org.springframework.boot' version "$springBootVersion"
+    id 'io.spring.dependency-management' version "$dependencyManagementVersion"
+    id 'groovy'
+}
+
+group = 'me.svdev'
```
Comment (ilkin-aliyev) [inline build.gradle]: change to related with ingress

File: src/main/java/me/svdev/mscategory/controller/InternalCategoryController.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+import static lombok.AccessLevel.PRIVATE;
+import static org.springframework.http.HttpStatus.CREATED;
+import static org.springframework.http.HttpStatus.NO_CONTENT;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("${root.internal}/v1/categories")
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/controller/InternalCategoryController.java]: change to direct path in annotation

File: src/main/java/me/svdev/mscategory/controller/SubcategoryController.java [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("v1/subcategories")
+@FieldDefaults(level = PRIVATE, makeFinal = true)
+public class SubcategoryController {
+    SubcategoryService subcategoryService;
+
+    @GetMapping("{id}")
+    public SubcategoryResponse validate(@PathVariable Long id) {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/controller/SubcategoryController.java]: rename to getSubcategory

File: src/main/java/me/svdev/mscategory/dao/entity/CategoryEntity.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+    Boolean isFavorite = false;
+
+    @CreationTimestamp
+    LocalDateTime updatedAt;
+
+    @UpdateTimestamp
+    LocalDateTime createdAt;
+
+    @OneToMany(fetch = EAGER, cascade = PERSIST, mappedBy = "category")
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/entity/CategoryEntity.java]: 1. change fetch type to LAZY

File: src/main/java/me/svdev/mscategory/dao/repository/CategoryRepository.java [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.lang.NonNullApi;
+
+import java.util.Optional;
+
+import static org.springframework.data.jpa.repository.EntityGraph.EntityGraphType.LOAD;
+
+public interface CategoryRepository extends JpaRepository<CategoryEntity, Long> {
+    @EntityGraph(value = "categories", type = LOAD)
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/dao/repository/CategoryRepository.java]: you can remove LOAD

File: src/main/java/me/svdev/mscategory/exception/ErrorHandler.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+import static org.springframework.http.HttpStatus.NOT_FOUND;
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
+
+    @ExceptionHandler(Exception.class)
+    @ResponseStatus(INTERNAL_SERVER_ERROR)
+    public ErrorResponse unexpectedErrorHandler(Exception exception) {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/exception/ErrorHandler.java]: 1. use method overloading
2. change log messages
3. don't use hardcoded exception messages except Exception.class

File: src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+
+import me.svdev.mscategory.dao.entity.CategoryEntity;
+import me.svdev.mscategory.model.request.CategoryRequest;
+import me.svdev.mscategory.model.response.CategoryResponse;
+
+import static java.util.Optional.ofNullable;
+
+public enum CategoryMapper {
+    CATEGORY_MAPPER;
+    public static CategoryEntity mapRequestToEntity(CategoryRequest categoryRequest) {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java]: remove static from everywhere

File: src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+        return CategoryEntity.builder()
+                .name(categoryRequest.getName())
+                .isFavorite(categoryRequest.getIsFavorite())
+                .isVisible(true)
+                .build();
+    }
+
+    public static CategoryResponse mapEntityToResponse(CategoryEntity categoryEntity) {
+        return new CategoryResponse(categoryEntity.getId(), categoryEntity.getName(), categoryEntity.getIsFavorite(), null);
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/mapper/CategoryMapper.java]: bad practice to return null, use builder pattern or different classes for building response

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+@FieldDefaults(level = PRIVATE, makeFinal = true)
+public class CategoryService {
+
+    CategoryRepository categoryRepository;
+
+    public CategoriesResponse getCategories() {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: use AOP for logging

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        log.info("ActionLog.getCategories.start");
+        var categories = new CategoriesResponse(categoryRepository.findAll().stream().map(CategoryMapper::mapEntityToResponse).toList());
+        log.info("ActionLog.getCategories.end");
+        return categories;
+    }
+
+    @Transactional(rollbackFor = {RuntimeException.class})
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: 1. Remove @Transactional
2. rollbackFor = RuntimeException.class this is invalid argument

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+    public CategoryResponse getCategory(Long id) {
+        log.info("ActionLog.getCategory.start");
+        var category = getCategoryIfExists(id);
+        log.info("ActionLog.getCategory.end");
+        return new CategoryResponse(category.getId(), category.getName(), category.getIsFavorite(), category.getSubcategories().stream().map(SubcategoryMapper::mapEntityToDto).toList());
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: use mapper for this

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+        log.info("ActionLog.deleteCategory.start");
+        var category = getCategoryIfExists(id);
+        category.setIsVisible(false);
+        log.info("ActionLog.deleteCategory.end");
+    }
+
+    @SneakyThrows
+    public CategoryEntity getCategoryIfExists(Long id) {
+        log.info("ActionLog.getCategoryIfExists.start");
+        var category = categoryRepository.findById(id).orElseThrow(() -> new NotFoundException(NOT_FOUND_ERROR_WITH_ID.getMessage() + id));
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: [Optional] - you can use .formatted instead +. concatenation

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+        log.info("ActionLog.saveCategory.start");
+        var category = CategoryMapper.mapRequestToEntity(categoryRequest);
+        categoryRepository.save(category);
+        log.info("ActionLog.saveCategory.end");
+    }
+
+    public void updateCategory(Long id, CategoryRequest categoryRequest) {
+        log.info("ActionLog.updateCategory.start");
+        var category = getCategoryIfExists(id);
+        CategoryMapper.partialUpdate(categoryRequest, category);
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: call repository.save everywhere, where not exist @Transactional

File: src/main/java/me/svdev/mscategory/service/CategoryService.java [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    public void deleteCategory(Long id) {
+        log.info("ActionLog.deleteCategory.start");
+        var category = getCategoryIfExists(id);
+        category.setIsVisible(false);
+        log.info("ActionLog.deleteCategory.end");
+    }
+
+    @SneakyThrows
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/CategoryService.java]: why you need it?

File: src/main/java/me/svdev/mscategory/service/SubcategoryService.java [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+
+import static lombok.AccessLevel.PRIVATE;
+import static me.svdev.mscategory.exception.ExceptionMessages.NOT_FOUND_ERROR_WITH_ID;
+import static me.svdev.mscategory.mapper.SubcategoryMapper.mapEntityToDto;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+@FieldDefaults(level = PRIVATE, makeFinal = true)
+public class SubcategoryService {
```
Comment (ilkin-aliyev) [inline src/main/java/me/svdev/mscategory/service/SubcategoryService.java]: refactor this service based on CategoryService

File: src/main/resources/application.yml [hunk 33]
```diff
@@ -0,0 +1,20 @@
+spring:
+  application:
+    name: ms.category
+  datasource:
+    url: jdbc:postgresql://${DB_HOST}:${DB_PORT}/${DB_NAME}
+    username: ${DB_USER}
+    password: ${DB_PASS}
+    driver-class-name: org.postgresql.Driver
+    hikari:
+      maximum-pool-size: 20
```
Comment (ilkin-aliyev) [inline src/main/resources/application.yml]: why you need it?

File: src/test/java/me/svdev/mscategory/MsCategoryApplicationTests.java [hunk 34]
```diff
@@ -0,0 +1 @@
+package me.svdev.mscategory;
```
Comment (ilkin-aliyev) [inline src/test/java/me/svdev/mscategory/MsCategoryApplicationTests.java]: add unit tests for all layers

File: src/main/java/az/ingress/mscategory/dao/entity/CategoryEntity.java [hunk 35]
```diff
@@ -0,0 +1,63 @@
+package az.ingress.mscategory.dao.entity;
+
+import lombok.*;
+import lombok.experimental.FieldDefaults;
+import org.hibernate.annotations.CreationTimestamp;
+import org.hibernate.annotations.UpdateTimestamp;
+import org.hibernate.annotations.Where;
+
+import javax.persistence.*;
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/dao/entity/CategoryEntity.java]: [Optional] - don't use * (wildcards for imports)

File: src/main/java/az/ingress/mscategory/exception/ErrorHandler.java [hunk 36]
```diff
@@ -0,0 +1,36 @@
...
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
+
+    @ExceptionHandler(Exception.class)
+    @ResponseStatus(INTERNAL_SERVER_ERROR)
+    public ErrorResponse errorHandler(Exception exception) {
+        log.error("ActionLog.errorHandler.error", exception);
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/exception/ErrorHandler.java]: 1.rename to handle 2.  in log write exception class name and stacktrace 3. use static import for enums

File: src/main/java/az/ingress/mscategory/exception/ErrorHandler.java [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+        return new ErrorResponse(exception.getMessage());
+    }
+
+    @ExceptionHandler(BadRequestException.class)
+    @ResponseStatus(BAD_REQUEST)
+    public ErrorResponse errorHandler(BadRequestException exception) {
+        log.error("ActionLog.errorHandler.error", exception);
+        return new ErrorResponse(exception.getMessage());
+    }
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/exception/ErrorHandler.java:44]: add error handling for 405 error

File: src/main/java/az/ingress/mscategory/exception/ExceptionMessages.java [hunk 38]
```diff
@@ -0,0 +1,18 @@
...
+
+import static lombok.AccessLevel.PRIVATE;
+
+@Getter
+@RequiredArgsConstructor
+@FieldDefaults(level = PRIVATE, makeFinal = true)
+public enum ExceptionMessages {
+    UNEXPECTED_ERROR("Unexpected error occurred!"),
+    NOT_FOUND_ERROR("Resource or data not found!"),
+    NOT_FOUND_ERROR_WITH_ID("Resource or data not found with ID: %d"),
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/exception/ExceptionMessages.java]: remove without id , and rename to NOT_FOUND_ERROR

File: src/main/java/az/ingress/mscategory/exception/ExceptionMessages.java [hunk 39]
```diff
@@ -0,0 +1,18 @@
...
+    BAD_REQUEST_ERROR("Bad request");
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/exception/ExceptionMessages.java]: remove this if this unused

File: src/main/java/az/ingress/mscategory/mapper/CategoryMapper.java [hunk 40]
```diff
@@ -0,0 +1,28 @@
...
+    public CategoryEntity mapRequestToEntity(CategoryRequest categoryRequest) {
+        return CategoryEntity.builder()
+                .name(categoryRequest.getName())
+                .isFavorite(categoryRequest.getIsFavorite())
+                .isVisible(true)
+                .build();
+    }
+
+    public CategoryResponse mapEntityToResponse(CategoryEntity categoryEntity) {
+        return new CategoryResponse(categoryEntity.getId(), categoryEntity.getName(), categoryEntity.getIsFavorite(), categoryEntity.getSubcategories().stream().map(SUBCATEGORY_MAPPER::mapEntityToResponse).toList());
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/mapper/CategoryMapper.java]: [Optional] - 1.separate lines by fields 2. Would be better use Builder

File: src/main/java/az/ingress/mscategory/service/SubcategoryService.java [hunk 41]
```diff
@@ -0,0 +1,53 @@
...
+
+    SubcategoryRepository subcategoryRepository;
+    CategoryService categoryService;
+
+    public SubcategoryResponse getSubcategory(Long id) {
+        return SUBCATEGORY_MAPPER.mapEntityToResponse(getSubcategoryIfExists(id));
+    }
+
+    public void saveSubcategory(SaveSubcategoryRequest saveSubcategoryRequest) {
+        var category = categoryService.getCategoryIfExists(saveSubcategoryRequest.getCategoryId());
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/service/SubcategoryService.java]: try to use getreferencebyid for saving child entities

File: src/main/java/az/ingress/mscategory/service/SubcategoryService.java [hunk 42]
```diff
@@ -0,0 +1,53 @@
...
+        var subcategory = SUBCATEGORY_MAPPER.mapRequestToEntity(saveSubcategoryRequest, category);
+        subcategoryRepository.save(subcategory);
+    }
+
+    public void updateSubcategory(@Log.Exclude Long id, UpdateSubcategoryRequest updateSubcategoryRequest) {
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/service/SubcategoryService.java]: remove @Log.Exclude

File: src/main/java/az/ingress/mscategory/service/SubcategoryService.java [hunk 43]
```diff
@@ -0,0 +1,53 @@
...
+
+    public void deleteSubcategory(Long id) {
+        var subcategory = getSubcategoryIfExists(id);
+        subcategory.setIsVisible(false);
+        subcategoryRepository.save(subcategory);
+    }
+
+    private SubcategoryEntity getSubcategoryIfExists(Long id) {
+        return subcategoryRepository.findById(id)
+                .orElseThrow(() -> new NotFoundException(NOT_FOUND_ERROR_WITH_ID.getMessage().formatted(id)));
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/service/SubcategoryService.java:51]:
```suggestion
                .orElseThrow(() -> new NotFoundException(NOT_FOUND_ERROR_WITH_ID.getMessage().formatted(id,"subcategory")));
```

File: src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy [hunk 44]
```diff
@@ -0,0 +1,89 @@
...
+
+class CategoryControllerTest extends Specification {
+    CategoryService categoryService
+    MockMvc mockMvc
+    def random = aNewEnhancedRandom()
+
+    def setup() {
+        categoryService = Mock()
+        def categoryController = new CategoryController(categoryService)
+        mockMvc = standaloneSetup(categoryController).build()
```
Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy]: add error handling

File: src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy [hunk 45]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    def "testGetCategories()"() {
+        given:
+        def subcategory = random.nextObject(SubcategoryResponse)
+        def category = new CategoryResponse(random.nextLong(), random.nextObject(String), random.nextBoolean(), List.of(subcategory))
```
Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy]: use groovy style change List.of to []

File: src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy [hunk 46]
```diff
@@ -0,0 +1,89 @@
...
+        def response = List.of(category)
+        def url = "/v1/categories"
+
+        def expectedResponse = """
+          [
+            {
+              "id": $category.id,
```
Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy:36]: Need to change values from dynamic to manual , for better testing

Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/controller/CategoryControllerTest.groovy:37]: nice 💯

File: src/test/groovy/az/ingress/mscategory/controller/InternalSubcategoryControllerTest.groovy [hunk 47]
```diff
@@ -0,0 +1,79 @@
...
+
+    def setup() {
+        subcategoryService = Mock()
+        def internalSubcategoryController = new InternalSubcategoryController(subcategoryService)
+        mockMvc = standaloneSetup(internalSubcategoryController).build()
+    }
+
+    def "testSaveSubcategory(saveSubcategoryRequest)"() {
+        given:
+        def request = new SaveSubcategoryRequest(10, "test", true)
```
Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/controller/InternalSubcategoryControllerTest.groovy]: use random

File: src/test/groovy/az/ingress/mscategory/service/CategoryServiceTest.groovy [hunk 48]
```diff
@@ -0,0 +1,144 @@
...
+        entity.isFavorite == request.isFavorite
+        entity.isVisible == true
+    }
+
+    def "testUpdateCategory(id, categoryRequest) success"() {
+        given:
+        def id = random.nextLong()
+        def request = random.nextObject(CategoryRequest)
+        def entity = random.nextObject(CategoryEntity)
+        CATEGORY_MAPPER.partialUpdate(request, entity)
```
Comment (ilkin-aliyev) [inline src/test/groovy/az/ingress/mscategory/service/CategoryServiceTest.groovy]: i think you don't need

File: src/main/java/az/ingress/mscategory/service/SubcategoryService.java [hunk 49]
```diff
@@ -0,0 +1,63 @@
...
+
+import static az.ingress.mscategory.exception.ExceptionMessages.SUBCATEGORY_ALREADY_EXISTS_ERROR_WITH_NAME;
+import static az.ingress.mscategory.exception.ExceptionMessages.SUBCATEGORY_NOT_EXISTS_ERROR_WITH_ID;
+import static lombok.AccessLevel.PRIVATE;
+import static az.ingress.mscategory.mapper.SubcategoryMapper.SUBCATEGORY_MAPPER;
+
+@Log
+@Service
+@RequiredArgsConstructor
+@Slf4j
```
Comment (ilkin-aliyev) [inline src/main/java/az/ingress/mscategory/service/SubcategoryService.java]: remove this

**Comment by ilkin-aliyev** [comment]
LGTM

**Review by ilkin-aliyev — APPROVED** [review]

**Pull request merged**